src/content/blog/val-vibes.mdx

```diff
@@ -0,0 +1,141 @@
+
+1. Indexing: generating embeddings for all public vals.
+2. Querying: generating an embedding for the search query, and comparing to the indexed embeddings.
+
+For indexing, we first create a Postgres table on Neon (using the [pgvector](https://github.com/pgvector/pgvector) extension):
+
+```
+CREATE TABLE vals_embeddings (id TEXT PRIMARY KEY, embedding VECTOR(1536));
+```
+
+Then we create a [cron val](https://docs.val.town/quickstarts/first-cron/) (or as I like to call them: inter-val…), which we set to run every hour. First we query all vals, and all existing embeddings, and make batches of 100 for the vals that we haven't indexed yet.
+
+<figure style={{ maxWidth: "700px" }}>
+  <Image src={Indexing1} alt="" />
+  <figcaption>
+  </figcaption>
+</figure>
+
+Then we fetch embeddings from OpenAI for each batch of 100 in parallel, and save them to Postgres.
+
+<figure style={{ maxWidth: "700px" }}>
+  <Image src={Indexing2} alt="" />
+  <figcaption>
+  </figcaption>
+</figure>
+
+This works! Here are the logs from a recent incremental run:
+
+<figure style={{ maxWidth: "700px" }}>
+  <Image src={Indexing3} alt="" />
+  <figcaption>
+  </figcaption>
+</figure>
+
+Querying is simpler still. We turn the query string into an embedding, and sort by similarity:
+
+<figure style={{ maxWidth: "700px" }}>
+  <Image src={Search1} alt="" />
+  <figcaption>
+  </figcaption>
+</figure>
+
+For a nicer UI, I forked an HTTP val (also by <a href="https://www.val.town/u/pomdtr">Achille Lacoin</a>) and hooked it up to my search function:
+
+<figure style={{ maxWidth: "700px" }}>
+  <Image src={Search2} alt="" />
+  <figcaption>
+  </figcaption>
+</figure>
+
+This works pretty well! For better ranking, we should probably include other signals, such as popularity, exact matches, and more, but even without that, this proved useful! We have since integrated this into the search of Val Town properly.
+
+### Blobs
+
+As promised, here is the implementation using blob storage. This doesn't require any sort of database, just a place to store bytes.
+
+I used Val Town's standard library, which contains [blob storage functions](https://docs.val.town/std/blob/). You can use it to store binary data, or JSON (which gets serialized/deserialized). So we can simply store a blob for every batch of 100 vals, and keep a separate JSON storage to keep track of which embedding is stored where:
+
+<figure style={{ maxWidth: "700px" }}>
+  <Image src={Blobindex} alt="" />
+  <figcaption>
+  </figcaption>
+</figure>
+
+Then for querying, we load all the blobs in memory, and compare them all against the embedding from the query string. This is of course not as efficient as using an indexed optimized for this sort of sorting, but not too bad for our (still) small number of vals.
+
+<figure style={{ maxWidth: "700px" }}>
+  <Image src={Blobsearch} alt="" />
+  <figcaption>
+  </figcaption>
+</figure>
+
+### Conclusion
```

Comment:
I'd add a section about productionzing it and where it lives now in the product and link people to some results and maybe have a screenshot of it

It's a great spot to talk about the challenges of making the prototype polished

And future directions, how we'll combine exact and vibes search into one list (which you talk about in linear)